### PR TITLE
river: add module

### PR DESCRIPTION
**Pull request by GaetanLepage**
### Description

Add the `wayland.windowManager.river` module to configure the [`river` wayland compositor](https://github.com/riverwm/river)

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [x] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

**Comment by JustinLovinger** [comment]
Environment variables do not seem to be set for systemd as documentation implies:

```
 wayland.windowManager.river.systemdIntegration
     Whether to enable river-session.target on river startup. This links to graphical-session.target. Some
     important environment variables will be imported to systemd and dbus user environment before reaching the
     target, including

     •   DISPLAY

     •   WAYLAND_DISPLAY

     •   XDG_CURRENT_DESKTOP
```

**Comment by berbiche** [comment]
Good work.

The module is complex and I think it could be made much simpler with a different design. 

River's config is basically a shell script where `riverctl` or `rivertile` is called with different sets of attributes.
The configuration could thus be represented as a nested attribute set where the leaves are the configuration value and the keys represent the path of the value to set.
This would remove a lot of boilerplate from the code generation and make it simpler to maintain if `riverctl` adds support for new commands (it would basically require no changes).

I've got the following script which seems to work:

``` nix
with lib;
let
  toValue = val:
    if isString val || isDerivation val then ''"${val}"'' 
    else if true == val then ''"enabled"''
    else if false == val then ''"disabled"''
    else if isInt val then toString val
    else if isFloat val then strings.floatToString val
    else if isList val then concatMapStringsSep " " toValue val
    else abort "unsupported type ${builtins.typeOf val}";

  # toCommand :: [string] -> attrs -> [string]
  toCommand = basePath: attrs:
   concatLists (mapAttrsToList (key: value:
      let
        path = basePath ++ [key];
        stringValue = concatStringsSep " " (path ++ [(toValue value)]);
        finalValue =
          if isAttrs value
          then toCommand path value
          else [stringValue];
      in finalValue
    ) attrs);
in
  toCommand
```

Here's an example with the nix expression and the expected generated commands.

```nix
toCommand ["riverctl"] {
  input.pointer-foo-bar = {
    accel-profile = "flat";
    pointer-accel = -0.3;
    events = "enabled";
    tap = "disabled";
  };
}
```
results in
```
riverctl input pointer-foo-bar accel-profile "flat"
riverctl input pointer-foo-bar pointer-accel -0.3
riverctl input pointer-foo-bar events "enabled"
riverctl input pointer-foo-bar tap "disabled"
```

Lists of primitive types would be supported but not lists of anything else. 

The caveat of the suggested syntax is that it won't support something that produces a result like this without more complex logic:

```
riverctl foo bar
riverctl foo bar baz
```

File: modules/services/window-managers/river/default.nix [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+      ${inputs}
+
+      ### EXTRA CONFIGURATION ###
+      ${cfg.extraConfig}
+
+      ### STARTUP PROGRAMS ###
+      ${concatMapStringsSep "\n" formatStartupProgram cfg.startupPrograms}
+
+      ### LAYOUT ###
+      riverctl default-layout ${cfg.config.layoutGenerator.name}
```
Comment (JustinLovinger) [inline modules/services/window-managers/river/default.nix:204]: Note, the name of the layout is not guaranteed to be equal to the filename of the executable, see <https://github.com/riverwm/river/wiki#my-windows-are-not-tiled>.

Comment (GaetanLepage) [reply modules/services/window-managers/river/default.nix:204]: Thanks for this remark !

**Comment by GaetanLepage** [comment]
Thank you very much for the feedback @berbiche and @JustinLovinger !
I am working on a rewrite and will push it once it will be ready !

File: modules/services/window-managers/river/default.nix [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+      ${cfg.config.layoutGenerator.name} ${cfg.config.layoutGenerator.arguments} &
```
Comment (JustinLovinger) [inline modules/services/window-managers/river/default.nix:205]: systemd integration should follow:

```
### SYSTEMD ###
${pkgs.dbus}/bin/dbus-update-activation-environment --systemd DISPLAY WAYLAND_DISPLAY XDG_CURRENT_DESKTOP XDG_SESSION_TYPE XDG_VTNR NIXOS_OZONE_WL
systemctl --user --no-block start river-session.target
```

with a conditional for `cfg.systemdIntegration` of course.

I based this on the Sway systemd integration, <https://github.com/nix-community/home-manager/blob/0ee5ab611dc1fbb5180bd7d88d2aeb7841a4d179/modules/services/window-managers/i3-sway/sway.nix#L320>, and tested it on my machine.

Comment (berbiche) [reply modules/services/window-managers/river/default.nix:205]: We should probably make that bit generic across Wayland window managers in a different PR

**Comment by GaetanLepage** [comment]
@JustinLovinger @berbiche, thanks to your very helpful comments, I have been able to set up a new implementation inspired by the [`toCommand` helper](https://github.com/nix-community/home-manager/pull/3977#issuecomment-1605882905) which I changed a little bit.

I will be glad to hear what you think about it now.

File: modules/services/window-managers/river/default.nix (hunk 1 above)
Comment (GaetanLepage) [reply modules/services/window-managers/river/default.nix]: I have created an option to eventually specify a name for the executable if it is not the same as the layout generator itself.

File: modules/services/window-managers/river/default.nix (hunk 2 above)
Comment (GaetanLepage) [reply modules/services/window-managers/river/default.nix]: I added the proper systemd `dbus` call.

File: modules/services/window-managers/river/options.nix [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+
+        whenTyping = mkOption {
+          type = types.bool;
+          default = false;
+          description = ''
+            Hide the cursor when pressing any non-modifier key.
+            Show the cursor again on any movement.
+          '';
+        };
+      };
```
Comment (berbiche) [inline modules/services/window-managers/river/options.nix:254]: IMO all these camelCase options are redundant.

There is no advantage having these options documented in Home Manager's options page over the official documentation.

**Review by berbiche — CHANGES_REQUESTED** [review]
Thanks for your hard work but the module is still too complex.

I believe the design I recommended in my initial comment is still much simpler.

The set of settings could be described in less than 30 lines.

``` nix
let
  primitives = types.oneOf [ types.str types.int types.float ... ];
  settingsType = types.attrsOf (types.attrsOf primitives)
in
{
  settings = mkOption {
    type = types.submodule {
      freeformType = settingsType;
      # An option that is explicitly described and typed, if required
      options.something = ...;
    };
  };

  # Same thing I guess
  # If it can be nested under `settings` in some way it would be great
  layouts = ...;
}
```

and in the config section, you could generate the configuration file with `toCommand "riverctl" cfg.settings`.

1. Does the order of settings matter in the configuration?

If yes, you can probably just call `toCommand` per key/settings in the order they should appear in the final configuration as nix stores attribute sets ordered by the key name.

**Comment by JustinLovinger** [comment]
@GaetanLepage, To add on to berbiche, take a look at RFC 42, <https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md>. The style of module you are writing is deprecated, and Nix is moving towards the style of module berbiche suggested.

**Comment by GaetanLepage** [comment]
@berbiche @justinlovinger Hi !
After several months I decided to come back to this module.
Indeed, adopting the RFC42-style was a better idea.

Please, tell me if the module better suits your expectation now :)

File: modules/services/window-managers/river.nix [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+            int
+            float
+            str
+            path
+            (attrsOf valueType)
+            (listOf valueType)
+          ]) // {
+            description = "River configuration value";
+          };
+        in valueType;
```
Comment (GaetanLepage) [inline modules/services/window-managers/river.nix:133]: In the end, I haven't felt that it was necessary to use a `types.submodule` with `freeformType`.
Indeed, I haven't found necessary to explicitly declare any option in this case.

By the way, this is the same thing that is done in the [hyprland module](https://github.com/nix-community/home-manager/blob/3d6791b3897b526c82920a2ab5f61d71985b3cf8/modules/services/window-managers/hyprland.nix#L113-L127).

File: modules/services/window-managers/river.nix [hunk 5]
```diff
@@ -0,0 +1,213 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
```
Comment (drupol) [inline modules/services/window-managers/river.nix:3]: This is deprecated. Just inherit only what you need out of it.

Comment (GaetanLepage) [reply modules/services/window-managers/river.nix]: Done :)

File: modules/services/window-managers/river.nix [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+          - `WAYLAND_DISPLAY`
+          - `XDG_CURRENT_DESKTOP`
+          - `NIXOS_OZONE_WL`
+          - `XCURSOR_THEME`
+          - `XCURSOR_SIZE`
+        '';
+      };
+
+      variables = lib.mkOption {
+        type = with types; listOf str;
```
Comment (drupol) [inline modules/services/window-managers/river.nix:85]: Most of those with are not needed, how about using types.listOf instead?

File: modules/services/window-managers/river.nix (hunk 5 above)
Comment (drupol) [reply modules/services/window-managers/river.nix]: Merci

File: modules/services/window-managers/river.nix (hunk 6 above)
Comment (GaetanLepage) [reply modules/services/window-managers/river.nix:85]: You would rather have `type = types.listOf types.str;` ?

Comment (drupol) [reply modules/services/window-managers/river.nix:85]: Yeah I personally prefer avoid using with in some cases.
I have to admit that it is an opinionated feedback and I wouldn't mind if you ignore it.

Comment (GaetanLepage) [reply modules/services/window-managers/river.nix]: I am fine with this as well.
I have only left the one for the complex `valueType` (`settings` option). Is that fine ?

Comment (drupol) [reply modules/services/window-managers/river.nix]: Yeah I think it's cool now, thanks 👍

**Comment by GaetanLepage** [comment]
Thanks for the review @drupol !

File: modules/services/window-managers/river.nix [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+    };
+
+    extraConfig = lib.mkOption {
+      type = types.lines;
+      default = "";
+      example = ''
+        rivertile -view-padding 6 -outer-padding 6 &
+      '';
+      description =
+        "Extra lines appended to <filename>$XDG_CONFIG_HOME/river/init</filename>";
```
Comment (rycee) [inline modules/services/window-managers/river.nix:163]:
```suggestion
        "Extra lines appended to {file}`$XDG_CONFIG_HOME/river/init`";
```

File: modules/services/window-managers/river.nix [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+    };
+
+    xwayland.enable = lib.mkEnableOption "XWayland" // { default = true; };
+
+    systemd = {
+      enable = lib.mkEnableOption null // {
+        default = true;
+        description = ''
+          Whether to enable {file}`river-session.target` on
+          river startup. This links to `graphical-session.target`.
```
Comment (rycee) [inline modules/services/window-managers/river.nix:72]:
```suggestion
          river startup. This links to {file}`graphical-session.target`.
```

File: tests/modules/services/window-managers/river/configuration.nix [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    test.stubs.river = { };
+
+    nmt.script = ''
+      riverInit=home-files/.config/river/init
+      assertFileExists "$riverInit"
+      assertFileIsExecutable "$riverInit"
+
+      normalizedConfig=$(normalizeStorePaths "$riverInit")
+      referenceConfig=$(normalizeStorePaths "${
+        pkgs.writeShellScript "river-init-expected" (readFile ./init)
```
Comment (rycee) [inline tests/modules/services/window-managers/river/configuration.nix:93]: Isn't the `writeShellScript` just adding the shebang? If so I think you could do
```suggestion
        ./init
```
and add the normalized shebang to the init file.

File: tests/modules/services/window-managers/river/configuration.nix [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+      };
+
+      extraConfig = ''
+        rivertile -view-padding 6 -outer-padding 6 &
+        some
+        extra config
+      '';
+    };
+
+    test.stubs.river = { };
```
[the remark this answers is not in the document]
Comment (GaetanLepage) [reply tests/modules/services/window-managers/river/configuration.nix]: Done

File: modules/services/window-managers/river.nix (hunk 7 above)
Comment (GaetanLepage) [reply modules/services/window-managers/river.nix]: Done

File: modules/services/window-managers/river.nix (hunk 8 above)
Comment (GaetanLepage) [reply modules/services/window-managers/river.nix]: Done

**Comment by rycee** [comment]
Thanks! Merged to master now 🙂

**Pull request merged**